src/typedefs.jl

```diff
@@ -29,28 +30,51 @@ end
 
 ## Values
 
-struct CategoricalValue{T, R <: Integer} <: AbstractString
+"""
+Default implementation of "categorical value" concept
+that references value of type `T`.
+"""
+struct CategoricalValue{T, R <: Integer}
     level::R
     pool::CategoricalPool{T, R, CategoricalValue{T, R}}
 end
 
+"""
+`String` categorical value.
+Provides `AbstractString` interoperability.
+"""
+struct CategoricalString{R <: Integer} <: AbstractString
+    level::R
+    pool::CategoricalPool{String, R, CategoricalString{R}}
+end
+
+# alias for all implementations of categorical value concept
+const CatValue = Union{CategoricalValue, CategoricalString}
+
 ## Arrays
 
-abstract type AbstractCategoricalArray{T, N, R, V, U} <: AbstractArray{Union{CategoricalValue{V, R}, U}, N} end
-AbstractCategoricalVector{T, R, V, U} = AbstractCategoricalArray{T, 1, R, V, U}
-AbstractCategoricalMatrix{T, R, V, U} = AbstractCategoricalArray{T, 2, R, V, U}
+# Type params:
+# * `T` original type of elements before categorization, could be nullable
+# * `N` array dimension
+# * `R` integral type for referncing category levels
+# * `V` orignial type of non-nullable elements before categorization
+# * `C` categorical value type that implements "categorical value" concept
+# * `U` type of null value, `Union{}` if the data is non-nullable
+abstract type AbstractCategoricalArray{T, N, R, V, C, U} <: AbstractArray{Union{C, U}, N} end
+AbstractCategoricalVector{T, R, V, C, U} = AbstractCategoricalArray{T, 1, R, V, C, U}
+AbstractCategoricalMatrix{T, R, V, C, U} = AbstractCategoricalArray{T, 2, R, V, C, U}
 
-struct CategoricalArray{T, N, R <: Integer, V, U} <: AbstractCategoricalArray{T, N, R, V, U}
+struct CategoricalArray{T, N, R <: Integer, V, C, U} <: AbstractCategoricalArray{T, N, R, V, C, U}
     refs::Array{R, N}
-    pool::CategoricalPool{V, R, CategoricalValue{V, R}}
+    pool::CategoricalPool{V, R, C}
```

Comment:
That would be nice, but `CategoricalPool{T,R,V}` is explicitly using `T` in `index`, `invindex` and `levels`.
`T` is indeed extractable from `V`, but I don't know how to void using it when defining `struct`.

Reply:
Probably it's possible to simplify `CategoricalPool` constructors by providing only `C`, but that might make it harder to create standalone pools.

Reply:
> T is indeed extractable from V, but I don't know how to void using it when defining struct.

What do you mean?

I wouldn't worry about how hard it is to create a `CategoricalPool`, currently isn't not even exported. We could create convenience constructors if we want.

Reply:
"avoid", if that caused confusion

By "dropping parameter V" have you meant from the `struct` or from the constructors?

Reply:
From the `struct`. You can provide convenience constructors which guess `V` from the input parameters if that's useful.

Reply:
Have you tried this?

Reply:
I'm not sure I understand what you envision to do.
`V` is used in multiple fields of the `struct CategoricalPool`, so either it stays as the type parameter of the `struct` or the `struct` layout is changed, although I'm not sure it's feasible.
If you actually propose to simplify the `CategoricalPool` ctors type parameters, I can look into that.

Reply:
Ah, too bad. Looks like we'll have to live with all these semi-redundant type parameters then...

Maybe just rename the type parameters so that `V` stands for "value" for consistency with `CategoricalPool`. You could replace `V` with `L` for "levels". Though if that causes too much churn we can do that later.